middle_end/flambda2.0/terms/flambda_primitive.ml

```diff
@@ -793,6 +793,11 @@ let effects_and_coeffects_of_unary_primitive p =
       (* [Obj.truncate] has now been removed. *)
       Effects.Only_generative_effects destination_mutability,
         Coeffects.No_coeffects
+    | Immutable_unique ->
+      (* CR vlaviron: this should never occur, but it's hard to express it
+         without duplicating the mutability type *)
+      Effects.Only_generative_effects destination_mutability,
+        Coeffects.No_coeffects
```

Comment:
In that case, we'd want to `Misc.fatal_error` I think: it corresponds exactly to a situation where an `Immutable_unique` value is duplicated. I guess this could eventually occur if/when some immutable but unique objects are updated ()e.g. record update) and the update process first copies it before patching the fields of the copied value, but it would be nice to have the assertion somewhere that an immutable but unique value is not duplicated.

Reply:
That's not the right place to make such an assertion. In theory there's nothing wrong with creating a new value from a copy of a unique value. If we wanted to add an assertion, that would be in closure_conversion, when we detect an unexpected input.